zcash_client_sqlite/src/wallet.rs

```diff
@@ -288,22 +288,39 @@ pub(crate) fn seed_matches_derived_account<P: consensus::Parameters>(
 
     // Keys are not comparable with `Eq`, but addresses are, so we derive what should
     // be equivalent addresses for each key and use those to check for key equality.
-    let uivk_match =
-        match UnifiedSpendingKey::from_seed(params, &seed.expose_secret()[..], account_index) {
-            // If we can't derive a USK from the given seed with the account's ZIP 32
-            // account index, then we immediately know the UIVK won't match because wallet
-            // accounts are required to have a known UIVK.
-            Err(_) => false,
-            Ok(usk) => UnifiedAddressRequest::all().map_or(
-                Ok::<_, SqliteClientError>(false),
-                |ua_request| {
-                    Ok(usk
-                        .to_unified_full_viewing_key()
-                        .default_address(ua_request)?
-                        == uivk.default_address(ua_request)?)
-                },
-            )?,
-        };
+    let uivk_match = {
+        let usk = UnifiedSpendingKey::from_seed(params, &seed.expose_secret()[..], account_index)
+            .map_err(|_| SqliteClientError::KeyDerivationError(account_index))?;
+
+        let (seed_addr, _) = usk.to_unified_full_viewing_key().default_address(Some(
+            UnifiedAddressRequest::all().expect("At least one supported pool feature ie enabled."),
+        ))?;
+
+        let (uivk_addr, _) = uivk.default_address(None)?;
+
+        #[cfg(not(feature = "orchard"))]
+        let orchard_match = false;
+        #[cfg(feature = "orchard")]
+        let orchard_match = seed_addr
+            .orchard()
+            .zip(uivk_addr.orchard())
+            .map(|(a, b)| a == b)
+            == Some(true);
+
+        let sapling_match = seed_addr
+            .sapling()
+            .zip(uivk_addr.sapling())
+            .map(|(a, b)| a == b)
+            == Some(true);
+
+        let p2pkh_match = seed_addr
+            .transparent()
+            .zip(uivk_addr.transparent())
+            .map(|(a, b)| a == b)
+            == Some(true);
+
+        orchard_match || sapling_match || p2pkh_match
+    };
```

Comment:
This is pretty much the same logic needed to fix #1490 (see the TODO in `add_account`). Does it make sense to fix that at the same time?